Add methods to clear stats

kpp opened the pull request.

Commit:
Add MeterRegistry::Clear method

Commit:
Add Meter::Reset method

Commit:
Add a test for Meter::Reset

Commit:
Add a test for MeterRegistry::Clear

Stebalien commented on meter.go:
```diff
@@ -53,6 +53,16 @@ func (m *Meter) Snapshot() Snapshot {
+// Clear sets accumulator, total and rate to zero and last update to current time.
+func (m *Meter) Clear() {
+	globalSweeper.snapshotMu.RLock()
+	defer globalSweeper.snapshotMu.RUnlock()
```
Needs to be a write lock.

Stebalien commented on meter.go:
```diff
@@ -53,6 +53,16 @@ func (m *Meter) Snapshot() Snapshot {
+func (m *Meter) Clear() {
+	globalSweeper.snapshotMu.RLock()
+	defer globalSweeper.snapshotMu.RUnlock()
+	atomic.StoreUint64(&m.accumulator, 0)
```
Unfortunately, this won't work:

1. Thread A calls `Mark`, transitioning the meter from idle to active. The meter is now written to the sweeper's registration channel.
2. Thread B calls `Clear`, resetting the accumulator to 0.
3. Thread A calls `Mark`, registering the meter _again_.

The meter will be registered twice.

A simple solution is to deduplicate meters inside the sweeper but I was trying to avoid keeping a map inside the sweeper to save memory. On the other hand, this map will be linear in the number of active meters, not linear in the number of total meters, so it's not _that_ bad.

kpp replied:
How about storing the info in the Meter itself whether it was registered or not?

Stebalien replied:
Yeah, that's probably fine. it's an extra word per meter but not the end of the world.

kpp replied:
I managed to write a reproducible race test for this case and pushed a fix.

kpp replied:
Done

Stebalien commented on meter.go:
```diff
@@ -53,6 +53,16 @@ func (m *Meter) Snapshot() Snapshot {
 }
 
+// Clear sets accumulator, total and rate to zero and last update to current time.
+func (m *Meter) Clear() {
```
Nit: I'd call this "reset". You "reset" a meter, you "clear" a meter registry.

kpp replied:
Right

Stebalien reviewed (changes requested):
Thanks!

Commit:
Fix data race in Meter::Reset

Commit:
Prevent Meter from being added twice

Stebalien committed:
nit: simplify conditional

Stebalien commented on sweeper.go:
```diff
@@ -113,6 +113,7 @@ func (sw *sweeper) update() {
 
 		// Mark this as idle by zeroing the accumulator.
 		swappedTotal := atomic.SwapUint64(&m.accumulator, 0)
+		atomic.StoreInt32(&m.registered, 0)
```
This will leave "registered" as 0 even if the meter was "revived" (see below).

Stebalien commented on meter.go:
```diff
@@ -53,6 +59,16 @@ func (m *Meter) Snapshot() Snapshot {
+	defer globalSweeper.snapshotMu.Unlock()
+	atomic.StoreUint64(&m.accumulator, 0)
+	m.snapshot.Rate = 0
+	atomic.StoreUint64(&m.snapshot.Total, 0)
```
We shouldn't need this.

Stebalien replied:
Ah. I see, we're now setting the total from _outside_ the main loop. Hm.

Stebalien committed:
reduce lock contention and fix a potential race

1. Handle `registered` inside the main loop. This way we can avoid atomics.
2. Don't start calculating bandwidth for newly active meters until they've been
active for a round. This:
  1. Ensures we only write to the snapshot from within the main loop.
  2. Gives us a better estimated bandwidth usage.

Stebalien committed:
reduce reset critical section

1. The last update is when we last saw activity. We don't need to update it when
we reset the meter.
2. Defer isn't _quite_ as fast as it could be.

Stebalien committed:
avoid allocating for interface conversions on clear

Stebalien commented on meter.go:
```diff
@@ -32,17 +32,23 @@ func (s Snapshot) String() string {
-		// I'm the first one to bump this above 0.
-		// Register it.
-		globalSweeper.Register(m)
+	if count > 0 {
```
?

Stebalien commented:
@kpp I needed to so some experimentation so it was simpler to make changes directly than to leave comments in a review. Do my changes LGTY?

Commit:
Fix outdated comment for Meter::Reset

kpp commented:
Your solution looks even better than mine. LGTM

kpp commented:
Unfortunately your solution does not pass tests from go-libp2p-core/metrics. See https://github.com/libp2p/go-libp2p-core/pull/71/files and play with it.

```
~/ipfs/go-libp2p-core/metrics$ go test
--- FAIL: TestBandwidthCounter (5.00s)
    bandwidth_test.go:111: expected 2000.000000 (±140.000000), got 3999.427234
    bandwidth_test.go:112: expected 1000.000000 (±70.000000), got 1999.713617
    bandwidth_test.go:111: expected 2000.000000 (±140.000000), got 3999.427234
    bandwidth_test.go:112: expected 1000.000000 (±70.000000), got 1999.713617
    bandwidth_test.go:111: expected 2000.000000 (±140.000000), got 3999.427234
...
```

Whereas mine works fine. Don't forget about `replace` in go.mod

Stebalien commented:
The issue is that this version starts tracking bandwidth one second later. I'll update those tests.

Stebalien committed:
correctly set the initial total

The previous attempt actually make the problem worse:

1. It didn't update the "last update" time.
2. It updated the accumulator but then didn't set the total. That meant that the
_next_ update would count the bandwidth from two time periods.

This change also reverts the changes to the test (the test was right, the code
was wrong).

Stebalien commented:
Actually, the issue is that I implemented this wrong, misdiagnosed why the test was failing, then broke the tests :man_facepalming:. Fixed.
